engine/logic_monad.ml

```diff
@@ -96,6 +99,8 @@ struct
     let (e, info) = Exninfo.capture e in
     raise (e,info) ()
 
+  let get_debug_cmd = fun () -> (); fun () -> Pervasives.(!)forward_read_debug_cmd ()
```

Comment:
```suggestion
  let get_debug_cmd = fun () -> (); fun () -> forward_read_debug_cmd.contents ()
```
avoids the deprecated Pervasives.

Alternatively put the ref and getter before `!` is shadowed in the file.

Reply:
You shouldn't even put this in this sealed module. It should only be defined somewhere in the Ltac plugin.

Reply:
I've tried several variations, but I haven't been able to get the routine to compile if I move it into `tactic_debub.ml`:

I get this error for the following code:

```
File "plugins/ltac/tactic_debug.ml", lines 175-179, characters 4-7:
175 | ....fun () ->
176 |       (match Logic_monad.forward_read_debug_cmd.contents with
177 |       | Some f -> f
178 |       | None -> failwith "forward_read_debug_cmd")
179 |      ()
Error: This expression should not be a function, the expected type is
       string Logic_monad.NonLogical.t


module S =
struct
  open Logic_monad.NonLogical

  let get_debug_cmd : string t =
    fun () ->
      (match Logic_monad.forward_read_debug_cmd.contents with
      | Some f -> f
      | None -> failwith "forward_read_debug_cmd")
     ()

end

    (if is_coqIDE () then
      S.get_debug_cmd
    else
      Proofview.NonLogical.catch Proofview.NonLogical.read_line
      :
```

If I remove `: string t`, I get

```
File "plugins/ltac/tactic_debug.ml", line 193, characters 6-21:
193 |       S.get_debug_cmd
            ^^^^^^^^^^^^^^^
Error: This expression has type unit -> string
       but an expression was expected of type 'a Proofview.NonLogical.t
```

As for `forward_read_debug_cmd`, IIUC if it is defined in `tactic_debug.ml` in the plugin then it can't be accessed from `idetop.ml`, in which case it some other file that's not closely related.

Reply:
Use the `make` function, or wrap stuff with the monadic return. In any case, this should *not* belong to the low level module defining monadic primitives since it is 1. derivable from the other constructs and 2. Ltac specific. I don't know if there is a fully satisfying solution that does not involve linking idetop to the Ltac plugin, but at least put it a the highest possible point in the dependency chain, not the lowest one.

Reply:
Thanks.  I was modeling my routine on `read_line`, which doesn't use `make`.

> this should *not* belong to the low level module defining monadic primitives ...

As I noted though perhaps not so clearly in logic_monad.ml:

![image](https://user-images.githubusercontent.com/1253341/106042604-1f352380-6092-11eb-85c8-5a6290d09edf.png)

> I don't know if there is a fully satisfying solution that does not involve linking idetop to the Ltac plugin, but at least put it a the highest possible point in the dependency chain, not the lowest one.

It has to be below `tactic_debug`, which is likely to be fairly low level.  For the moment I put it in `Dyn`, which makes some sense since the plugins are dynamically loaded code.